rfcs/20190117-tf-module.md

```diff
@@ -0,0 +1,634 @@
+    properties of modules which are properties of this module (and so on).
+
+    >>> a = tf.Module()
+    >>> b = tf.Module()
+    >>> c = tf.Module()
+    >>> a.b = b
+    >>> b.c = c
+    >>> assert list(a.submodules) == [b, c]
+    >>> assert list(b.submodules) == [c]
+    >>> assert list(c.submodules) == []
+
+    Returns:
+      A generator over all submodules.
+    """
+
+  @property
+  def owned_submodules(self):
+    """Returns an iterator of immediate child modules.
+
+    Child modules are modules which are found as properties of the current
+    module.
+
+    >>> a = tf.Module()
+    >>> b = tf.Module()
+    >>> c = tf.Module()
+    >>> a.b = b
+    >>> b.c = c
+    >>> assert list(a.children) == [b]
+    >>> assert list(b.children) == [c]
+    >>> assert list(c.children) == []
+
+    Returns:
+      A generator over all child modules.
+    """
+
+class ModuleMetaclass(type):
+  def __new__(mcs, name, bases, clsdict):
+    """Modifies module subclasses when they are defined.
+
+    Wraps all methods and properties in clsdict such that enter the modules
+    name scope before evaluation.
+
+    >>> class MyModule(tf.Module):
+    ...   def __call__(self, x):
+    ...     if not hasattr(self, 'w'):
+    ...       self.w = tf.Variable(tf.zeros([x.shape[1], 64]), name='w')
+    ...     return tf.matmul(x, w)
+
+    Without any user intervention __call__ will enter the modules name scope,
+    allowing variables and submodules to be created. In some cases this is not
+    desirable, so there exists an opt out annotation(`@no_auto_name_scope`)
+    which instructs the metaclass to not wrap these functions.
+    """
+
+  def __call__(cls, *args, **kwargs):
+    """Creates new module instances.
+
+    To allow other modules/variables to be created in the constructor of a
+    module (and be named appropriately) we enter the module name scope inside
+    the base (Module) constructor. We close the name scope in the metaclass
+    before returning the new instance.
+
+    >>> class MyModule(tf.Module):
+    ...   def __init__(self):
+    ...     super(MyModule, self).__init__()
+    ...     # ^ Enters the module name scope but does not exit.
+    ...
+    ...     self.v = tf.Variable(1., name='v')
+    ...     # ^ Will be named "my_module/v"
+    ...
+    ...     self.m = MyOtherModule()
+    ...     # ^ Will be named "my_module/my_other_module"
+
+    >>> m = MyModule()
+    >>> # ^ Name scope will be closed thanks to metaclass.
+    """
+```
+
+### Naming
+
+Having human readable names for ops and variables is essential for debugging.
+Each `tf.Module` will have a name scope which is determined in the constructor.
+By default this will be a `lower_camel_case` version of the `ClassName`:
+
+```python
+class MyModule(tf.Module):
+  def __init__(self, name=None):
+    super(MyModule, self).__init__(name=name)
+    self.v = tf.Variable(1., name='v')
+
+>>> m = MyModule()
+>>> m.name
```

Comment:
What happens if the module name ends in '/'? Usually that is used to set _absolute_ name scopes ignoring the current stack. Similarly, what happens if a child module does it?

Reply:
In the experimental implementation of this I've enforced (in the constructor) that the module name is a valid Python 2 identifier [0], so `/` is not allowed.

[0] https://docs.python.org/2/reference/lexical_analysis.html#identifiers